reset surface damage if it gets unmapped

cmeissl opened the pull request:
~If the client attaches a new buffer (either NIL or some new buffer) we should reset the current damage or we will end up
with wrong damage in case of sync subsurfaces where multiple attach/commits are called on the surface without a parent commit in between.~

If a surface gets unmapped by attaching a `NULL`  buffer we should also clear the pending damage.

fixes #494

vberger commented:
Could you expand on how this fixes #494 ? Because this changes feels to me like it's not a correct thing to do, we are effectively discarding damage submitted by the client when doing that, we does not seem right.

Drakulix commented:
Right, this seems like, if a compositor missed some damage of a non-sync surface (maybe becaues it was not mapped on the active workspace or something like that), it will miss some damage.

cmeissl commented:
> Could you expand on how this fixes #494 ? Because this changes feels to me like it's not a correct thing to do, we are effectively discarding damage submitted by the client when doing that, we does not seem right.

The issue in #494 is that the surface has damage without any buffer attached. Then an unwrap on the buffer dimensions will panic. While this could also be solved by handling that in the code where unwrap is called I don't see a reason why we should keep damage on a surface when a NIL buffer has been attached. Also I don't think that it is correct to keep damage from another buffer as the damage should only be valid for the attached buffer?

> Right, this seems like, if a compositor missed some damage of a non-sync surface (maybe becaues it was not mapped on the active workspace or something like that), it will miss some damage.

I don't think this will miss some damage cause the "old" damage is only cleared if the client attached a new buffer or NIL, which implies the client also submitted new damage for the new buffer if any. Why should we keep damage from an old buffer or NIL for?

Drakulix commented:
> I don't think this will miss some damage cause the "old" damage is only cleared if the client attached a new buffer or NIL, which implies the client also submitted new damage for the new buffer if any. What for should we keep damage from an old buffer or NIL?

I don't think we want to keep damage, when a null-buffer is attached, because any new buffers will likely damage the whole surface anyway. (If anything, we would want to set damage to the size of the surface.)

But if downstream has not processed a buffer (so `mem::replace` returns `NewBuffer`), it might never have seen the damage. So if downstream renders the next buffer, it will fail to redraw all regions, that were damaged in the meantime. We need to accumulate damage, as long as clients send us valid buffers.

cmeissl commented:
From the spec:
> wl_surface.commit assigns pending damage as the current damage, and clears pending damage. The server will clear the current damage as it repaints the surface.

merge_into handles the commit for a sync subsurface if I have not misunderstood something, so this may apply or not. Not sure about that.

cmeissl commented:
> > I don't think this will miss some damage cause the "old" damage is only cleared if the client attached a new buffer or NIL, which implies the client also submitted new damage for the new buffer if any. What for should we keep damage from an old buffer or NIL?
> 
> I don't think we want to keep damage, when a null-buffer is attached, because any new buffers will likely damage the whole surface anyway. (If anything, we would want to set damage to the size of the surface.)
> 
> But if downstream has not processed a buffer (so `mem::replace` returns `NewBuffer`), it might never have seen the damage. So if downstream renders the next buffer, it will fail to redraw all regions, that were damaged in the meantime. We need to accumulate damage, as long as clients send us valid buffers.

I am still not convinced it is wrong to clear the damage. If a new buffer gets attached the client also has to provide the damage, so I can't see how that would make the compositor miss some damage. But maybe I miss something here.

Drakulix commented:
>  the client also has to provide the damage

The problem is "damage in relation to what?" and the answer is.. the last buffer/commit.
And because we can track, if downstream even has processed the last commit, we should accumulate for downstream.

So basically what I would suggest is:
```
if self.buffer.is_some() {
    if let Some(BufferAssignment::Removed) = &self.buffer {
        into.damage.clear();
    }
    if let Some(BufferAssignment::NewBuffer { buffer, .. }) =
        std::mem::replace(&mut into.buffer, self.buffer)
    {
        buffer.release();
    } else {
        into.damage.clear();
    }
}
```

cmeissl commented:
> > the client also has to provide the damage
> 
> The problem is "damage in relation to what?" and the answer is.. the last buffer/commit. And because we can track, if downstream even has processed the last commit, we should accumulate for downstream.
> 
> So basically what I would suggest is:
> 
> ```rust
> if self.buffer.is_some() {
>     if let Some(BufferAssignment::Removed) = &self.buffer {
>         into.damage.clear();
>     }
>     if let Some(BufferAssignment::NewBuffer { buffer, .. }) =
>         std::mem::replace(&mut into.buffer, self.buffer)
>     {
>         buffer.release();
>     } else {
>         into.damage.clear();
>     }
> }
> ```

Indeed, if there is already a buffer and damage it could miss some damage, thanks :+1: 
So it should be enough to clear the damage in case the buffer got explicitly removed only.
`std::mem::replace` returning `Some(BufferAssignment::Removed)` should already be handled with this change and
`None` should probably not contain damage at all?
